dbsnap_verify/__init__.py

```diff
@@ -0,0 +1,154 @@
+from time import time
+
+import datetime
+
+import json
+
+import boto3
+s3 = boto3.client("s3")
+
+import logging
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+
+secs_in_day = 86400
+
+def now_timestamp():
+    return int(time())
+
+def timestamp_to_datetime(timestamp):
+    return datetime.datetime.fromtimestamp(timestamp)
+
+def now_datetime():
+    return datetime.datetime.now()
+
+def today_date():
+    return now_datetime()
+
+def tomorrow_date():
+    return(now_datetime() + datetime.timedelta(days=1))
+
+def datetime_to_date_str(dt):
+    return dt.strftime("%Y-%m-%d")
+
+def date_str_to_datetime(date_str):
+    return datetime.datetime(*map(int, date_str.split("-")))
+
+def get_snapshot_description(event):
+    """if it exists return the snapshot definition, else None"""
+    return None
+
+def get_database_description(event):
+    """if it exists return the database definition, else None"""
+    return None
+
+def start_restore_from_snapshot(event):
+    pass
+
+def wait(state_doc):
+    logger.info("Looking for the {snapshot_date} snapshot of {database}".format(**state_doc))
+    description = get_snapshot_description(event)
+    if description:
+        restore(event, state_doc)
+    elif today_date() > date_str_to_datetime(state_doc["snapshot_date"]):
+        logger.warning("Alert! we never found the {snapshot_date} snapshot for {database}".format(**state_doc))
+        alarm("asdfasdfasdkfjnasdf naskdfn aksdjfn")
+    else:
+        set_state_doc_in_s3(state_doc, "wait")
+        logger.info("Did not find the {snapshot_date} snapshot of {database}".format(**state_doc))
+        logger.info("Going to sleep.")
+
+def restore(state_doc):
+    description = get_database_descripton(state_doc)
+    if description is None:
+        set_state_doc_in_s3(state_doc, "restore")
+        start_restore_from_snapshot(state_doc)
+    elif description[DBInstanceStatus] == "available":
+        set_state_doc_in_s3("verify")
+        verify(state_doc, description)
+        
+def verify(state_doc, description):
+    reset_master_password()
+    connection = connect_to_endpoint(description["endpoint"])
+    result = run_all_the_tests(connection, state_doc["verfication_checks"])
```

Comment:
If we do make this lambda connect to the database and run queries, it might be good to just have to run a function, like `SELECT is_restore_healthy();`. This'd make it easy for applications to determine what it means to be healthy from a restore, without much involvement from ops.

Reply:
like a stored procedure that does the testing/verifying?

Reply:
Right. This is what we do with pgbouncer for example, when a new connection is opened. PgBouncer just calls [`SELECT activerecord_setup()`](https://github.com/remind101/r101-api/blob/3e49ccb86aa34ca579f9c7527ee7dd11abf23854/db/structure.sql#L181-L183) which does all the necessary connection initialization that's needed.

Reply:
Cool idea.  I still think having the `SELECT is_restore_healthy();` should be passed using configuration and not hard coded.